esphome/components/pylontech/pylontech.h

```diff
@@ -0,0 +1,89 @@
+#pragma once
+
+#include "esphome/core/component.h"
+#include "esphome/core/defines.h"
+#include "esphome/components/uart/uart.h"
+
+#include "esphome/components/sensor/sensor.h"
+#ifdef USE_TEXT_SENSOR
+#include "esphome/components/text_sensor/text_sensor.h"
+#endif
+
+#ifndef PYLONTECH_NUM_BATTERIES
+// set by add_num_batteries() in __init__.py
+// this fallback is for linting and IDE stuff
+#define PYLONTECH_NUM_BATTERIES 3  // NOLINT
+#endif
+
+namespace esphome {
+namespace pylontech {
+
+static const uint8_t NUM_BUFFERS = 20;
+static const uint8_t TEXT_SENSOR_MAX_LEN = 8;
+
+#define PYLONTECH_ENTITY_(type, name) \
+ protected: \
+  type *name##_[PYLONTECH_NUM_BATTERIES]{}; /* NOLINT */ \
+\
+ public: \
+  void set_##name(type *name, int batnum) { /* NOLINT */ \
+    this->name##_[batnum - 1] = name; \
+  }
+
+#define PYLONTECH_SENSOR(name) PYLONTECH_ENTITY_(sensor::Sensor, name)
+#define PYLONTECH_TEXT_SENSOR(name) PYLONTECH_ENTITY_(text_sensor::TextSensor, name)
+
+class PylontechComponent : public PollingComponent, public uart::UARTDevice {
+ public:
+  PYLONTECH_SENSOR(voltage)
+  PYLONTECH_SENSOR(current)
+  PYLONTECH_SENSOR(temperature)
+  PYLONTECH_SENSOR(temperature_low)
+  PYLONTECH_SENSOR(temperature_high)
+  PYLONTECH_SENSOR(voltage_low)
+  PYLONTECH_SENSOR(voltage_high)
+
+  PYLONTECH_SENSOR(coulomb)
+  PYLONTECH_SENSOR(mos_temperature)
```

Comment:
All sensor code needs to be wrapped similar to text sensors.
```suggestion
#ifdef USE_SENSOR
  PYLONTECH_SENSOR(voltage)
  PYLONTECH_SENSOR(current)
  PYLONTECH_SENSOR(temperature)
  PYLONTECH_SENSOR(temperature_low)
  PYLONTECH_SENSOR(temperature_high)
  PYLONTECH_SENSOR(voltage_low)
  PYLONTECH_SENSOR(voltage_high)

  PYLONTECH_SENSOR(coulomb)
  PYLONTECH_SENSOR(mos_temperature)
#endif
```